respringless_tweaks.md

Luki120 opened the pull request:
* Closes #7

Luki120 committed:
create Assets folder

Luki120 committed:
create respringless_tweaks.md

Luki120 committed:
updated README

Luki120 committed:
update image

Luki120 committed:
update indentation & notification name

Luki120 committed:
better

Luki120 committed:
update subclass_wrapper.md image location

Luki120 committed:
update gitignore

Luki120 committed:
added the Respringless tweak project

cstanze commented:
Looks good to me, well done!

NightwindDev commented on respringless_tweaks.md:
```diff
@@ -0,0 +1,152 @@
+# How do you create a respringless tweak?
+
+* It's actually really easy if you understand the basics of it. What I'll make is a simple tweak that adds a blur to the LockScreen, with the option to change the its alpha/intensity on the fly. Think about it, how could that happen without a respring? Simple, Preferences needs to communicate with SpringBoard to notify that the blur value has been changed, so that it can refresh the UI. The `NSNo ...
```
```suggestion
*What we'll make is a simple tweak that adds a blur to the lock screen, with the option to change the its alpha/intensity on the fly. How can this happen without a respring? The Settings app will need to communicate with SpringBoard to notify that the blur value has been changed, so that it can refresh the UI. The `NSNotificationCenter` Foundation API makes that really straightforward.
```

NightwindDev commented on respringless_tweaks.md:
```diff
@@ -0,0 +1,152 @@
+# How do you create a respringless tweak?
+
+* It's actually really easy if you understand the basics of it. What I'll make is a simple tweak that adds a blur to the LockScreen, with the option to change the its alpha/intensity on the fly. Think about it, how could that happen without a respring? Simple, Preferences needs to communicate with SpringBoard to notify that the blur value has been changed, so that it can refresh the UI. The `NSNo ...
+I won't explain what `NSNotificationCenter` does or how it works in too much detail because if you've got this far then you should already know the basics of it, if not then I'd suggest reading its documentation.
```
```suggestion
You can explore the `NSNotificationCenter` API's documentation here: https://developer.apple.com/documentation/foundation/nsnotificationcenter.
```

NightwindDev commented on respringless_tweaks.md:
````diff
@@ -0,0 +1,152 @@
+Open your terminal and create a tweak project, your `Tweak.x` should look like this:
+
+```objc
+#import "Tweah.h"
````
```suggestion
#import "Tweak.h"
```

NightwindDev commented on respringless_tweaks.md:
```diff
@@ -0,0 +1,152 @@
+
+static void loadPrefs(void) {
+
+    NSUserDefaults *prefs = [[NSUserDefaults alloc] initWithSuiteName: kSuiteName];
```
```suggestion
    NSUserDefaults *prefs = [[NSUserDefaults alloc] initWithSuiteName:@"com.example.respringlesstweakprefs"];
```
Let's try to make the code as portable as possible.

NightwindDev commented on respringless_tweaks.md:
```diff
@@ -0,0 +1,152 @@
+
+}
+
+static _UIBackdropView *blurView;
```
This should probably be a `%property`, not a global static variable.

NightwindDev commented on respringless_tweaks.md:
```diff
@@ -0,0 +1,152 @@
+    %orig;
+    [self setupBlur];
+
+    [NSDistributedNotificationCenter.defaultCenter addObserver:self selector:@selector(updateBlurIntensity) name:RespringlessTweakDidUpdateBlurIntensityNotification object: nil];
```
Might want to show an alternative way of doing this which does not require a new selector, rather using a block.

NightwindDev commented on respringless_tweaks.md:
````diff
@@ -0,0 +1,152 @@
+
+```objc
+@import UIKit;
+#import "Common.h"
````
Might want to think about removing this file and adding it to the main `Tweak.h` for a more concise example project.

NightwindDev commented on respringless_tweaks.md:
````diff
@@ -0,0 +1,152 @@
+@end
+```
+
+Finally, `Common.h`
+
+```objc
+static NSString *const kSuiteName = @"me.luki.respringlesstweakprefs";
+
+static NSNotificationName const RespringlessTweakDidUpdateBlurIntensityNotification = @"RespringlessTweakDidUpdateBlurIntensityNotification";
+
+@interface NSDistributedNotificationCenter: NSNotificationCenter
+@end
+```
````
Probably should move all this to `Tweak.h`

NightwindDev commented on respringless_tweaks.md:
````diff
@@ -0,0 +1,152 @@
+```objc
+static NSString *const kSuiteName = @"me.luki.respringlesstweakprefs";
+
+static NSNotificationName const RespringlessTweakDidUpdateBlurIntensityNotification = @"RespringlessTweakDidUpdateBlurIntensityNotification";
````
Using a more specific notification name would be nicer, for example `com.example.respringlesstweakprefs/DidUpdateBlurIntensityNotification`. Also I think removing the constant is best

NightwindDev commented on respringless_tweaks.md:
````diff
@@ -0,0 +1,152 @@
+Finally, `Common.h`
+
+```objc
+static NSString *const kSuiteName = @"me.luki.respringlesstweakprefs";
````
```suggestion

```
Remove the `kSuiteName` constant

NightwindDev commented on respringless_tweaks.md:
```diff
@@ -0,0 +1,152 @@
+
+## Explanation
+
+* First we create the `loadPrefs(void)` function that will contain our prefs, adding the `float` key for the blur intensity.
```
```suggestion
* First we create the `void loadPrefs(void)` function that will contain our prefs, adding the `CGFloat` key for the blur intensity.
```
Specify the return type and use `CGFloat` since we are dealing with graphics

NightwindDev commented on Projects/Respringless/Common.h:
```diff
@@ -0,0 +1,6 @@
+static NSString *const kSuiteName = @"me.luki.respringlesstweakprefs";
+
+static NSNotificationName const RespringlessTweakDidUpdateBlurIntensityNotification = @"RespringlessTweakDidUpdateBlurIntensityNotification";
+
+@interface NSDistributedNotificationCenter: NSNotificationCenter
+@end
```
```suggestion
```
Remove this file

NightwindDev commented on Projects/Respringless/Prefs/Makefile:
```diff
@@ -0,0 +1,10 @@
+BUNDLE_NAME = Prefs
+
+Prefs_FILES = RTRootVC.m
+Prefs_CFLAGS = -fobjc-arc
+Prefs_FRAMEWORKS = UIKit
+Prefs_INSTALL_PATH = /Library/PreferenceBundles
+Prefs_PRIVATE_FRAMEWORKS = Preferences
+
+include $(THEOS)/makefiles/common.mk
+include $(THEOS_MAKE_PATH)/bundle.mk
```
```suggestion
BUNDLE_NAME = RespringlessPrefs

RespringlessPrefs_FILES = RTRootVC.m
RespringlessPrefs_CFLAGS = -fobjc-arc
RespringlessPrefs_FRAMEWORKS = UIKit
RespringlessPrefs_INSTALL_PATH = /Library/PreferenceBundles
RespringlessPrefs_PRIVATE_FRAMEWORKS = Preferences

include $(THEOS)/makefiles/common.mk
include $(THEOS_MAKE_PATH)/bundle.mk
```

NightwindDev commented on Projects/Respringless/Prefs/RTRootVC.m:
```diff
@@ -0,0 +1,22 @@
+	NSUserDefaults *prefs = [[NSUserDefaults alloc] initWithSuiteName: kSuiteName];
+	[prefs setObject:value forKey:specifier.properties[@"key"]];
+
+	[NSDistributedNotificationCenter.defaultCenter postNotificationName:RespringlessTweakDidUpdateBlurIntensityNotification object:nil];
```
```suggestion
	[NSDistributedNotificationCenter.defaultCenter postNotificationName:@"com.example.respringlesstweakprefs/DidUpdateBlurIntensityNotification" object:nil];
```
Might want to remove the constant for less resource files in the example

NightwindDev commented on Projects/Respringless/Prefs/Resources/Root.plist:
```diff
@@ -0,0 +1,33 @@
+			<key>default</key>
+			<real>0.85</real>
+			<key>defaults</key>
+			<string>me.luki.respringlesstweakprefs</string>
```
```suggestion
			<string>com.example.respringlesstweakprefs</string>
```

NightwindDev reviewed (changes requested):
Overall, quite good. Please take a look at the changes I proposed. The main goal here is to make the tweak as generalized as possible and to minimize the amount of files as well.

Please consider removing the `icon.png`, `icon@2x.png`, and `icon@3x.png` files (and any references to the icon) for a cleaner structure. Also, please change the `Prefs` subproject to something more unique - like `RespringlessPrefs`. 

In addition, I would add a check in the `-setPreferenceValue:specifier:` implementation to only post the notification if the specifier is the one that handles the blur. Otherwise, every time any setting gets changed in the view controller, a notification will get posted, which we do not want. The check does add a bit of verbosity, but when adding onto the example tweak, optimizations do need to be taken into account.

Luki120 committed:
rewording

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
rewording

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
rewording + applied suggested changes

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
Remove `Common.h`, renamed notification

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
renamed prefs subproject

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
Fixed typo

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
`CGFloat` > `float`

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
style change

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
fixed vc name in `Info.plist`

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
more generic names

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
control file changes

Co-authored-by: Nightwind <81449663+NightwindDev@users.noreply.github.com>

Luki120 committed:
removed references to the icon

Luki120 committed:
fixed compilation + runtime errors

Luki120 committed:
made `blurView` a property rather than a static var

Luki120 committed:
last changes
• added a check to only post the notification if the specifier is blurIntensity
• renamed prefs subproject

Luki120 commented:
Addressed all changes

Luki120 committed:
[respringless_tweaks.md] updated with latest code changes

NightwindDev reviewed (approved):
Looks good to me, thank you!
